src/openfermion/measurements/_bayesian_estimators.py

```diff
@@ -0,0 +1,845 @@
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+"""
+Classes for Bayesian estimators for quantum phase estimation.
+Estimators described in arXiv:1809.09697 (in particular in App.C and App.C.1)
+"""
+import warnings
+import numpy
+from numpy import cos, sin, pi
+from scipy import sparse, optimize
+from openfermion.config import EQ_TOLERANCE
+
+
+class FourierProbabilityDist(object):
+    """
+    Stores a multivariant Fourier representation of a periodic function:
+
+    f(phi_0,phi_1,...) = 
+        sum_j A_j sum_n (c_{j,2n} cos(n*phi_j) +
+                         c_{j,2n-1} sin(n*phi_j))
+
+    In particular, this class stores values of A_j, c_{j,2n},
+    c_{j,2n+1} from the above equation.
+
+    It also provides routines for multiplying f by functions of the form
+    cos^2(n*phi_j + beta), and for calculating maximum-likelihood
+    distributions of the A_j variables.
+
+    More details can be found in arXiv:1809.09697, Appendix C.
+    """
+
+    def __init__(self,
+                 amplitude_guess=None,
+                 amplitude_vars=None,
```

Comment:
Since this just represents a probability distribution, it doesn't make sense to talk about guesses and their variances. You should get rid of the `amplitude_vars` variable here; it doesn't look like you are using it anywhere anyway. `amplitude_guess` should be renamed because it's not a guess; it's just a specification of the coefficients which are part of the description of the probability distribution.

Reply:
So the variance is definitely a part of the distribution rather than a part of the estimator, and even though it's not currently used here it could well be later. I agree though that the word 'guess' here might be ill-defined. The reason why 'vector_guess' and 'amplitude_guess' exist is to allow for a reset function. This in turn is because creating these matrices is a fairly expensive procedure, so I'd rather not do that every time I want to perform an estimation.

Perhaps the solution is to pull _matrices out of FourierProbabilityDist? But then BayesEstimator is getting quite large.

Reply:
You should pull the matrix generation function out of FourierProbabilityDist because they aren't part of the description of the probability distribution. Don't stick them into BayesEstimator; just have them be stand-alone functions.

My understanding is that the probability distribution is completely defined by the A's and the c's, so I don't see where the variance is coming in. If you mean that the A's aren't known for sure, and we only have guesses for them, then we're actually representing a different object, which is more like a probability distribution over probability distributions.

Reply:
Hi Kevin, sorry for taking so long to get back to this.

The probability distribution is a distribution over both the amplitudes (A's) and the angles. For the angles I'm storing a representation of the entire function, but for the amplitudes I'm just storing an estimate of the average (as this is what's needed to update the angles), and data about the experiments (_p_vecs), and an initial estimate (amplitude_guess, amplitude_vars) to update this via maximum likelihood estimation.

I think pulling the matrix generation function out of FourierProbabilityDist is a great idea. I'm also going to pull out all of the estimation functions used in calculating the ML estimation out of BayesEstimator and leave them as separate functions in _probability_distributions.py , as they have a similar function.